### PR TITLE
Verify External IP is reachable or log a warning if not

### LINKED ISSUE #4418: Log a warning if not dialable
Especially when using flag --p2p-host-ip, users should be given a warning when one or more of their advertised multi-addresses are not accessible at start up.

Ideally, this would happen in the p2p start routine as a ping to all of the advertised multiaddresses. 
Additionally, this may be valuable to periodically check.

### PR DESCRIPTION
**Pull request by fabdarice**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
This PR will attempt at start-up to dial to the ip provided by **--p2p-host-ip** and ensure its connectivity. If this IP is not reachable, log a warning message to the user.

**Which issues(s) does this PR fix?**

Fixes #4418

**Other notes for review**
- Please note that attempting to connect to our own public address can result in false negative, typically if a user is behind a router and haven't configured properly port forwarding.
- I wasn't sure if I was supposed to use libp2p/host to verify connectivity to the public IP or if net.DialTimeout was enough

### CODE REVIEW

File: beacon-chain/p2p/service.go [hunk 1]
```diff
@@ -649,3 +654,20 @@ func logExternalDNSAddr(id peer.ID, addr string, port uint) {
 		).Info("Node started external p2p server")
 	}
 }
+
+// Attempt to dial an address to verify its connectivity
+func verifyExternalIPAddr(addr string, port uint) {
```
Comment (nisdas) [inline beacon-chain/p2p/service.go:659]: can you also add verification for the udp port provided ?

Comment (fabdarice) [reply beacon-chain/p2p/service.go:659]: added.

Comment (nisdas) [reply beacon-chain/p2p/service.go:659]: That is not what I meant, we have 2 different ports for both tcp and udp. So we also need to ensure they are publically reachable for both. You can access the set udp port in the service through `s.cfg.UDPPort`. Use that to also check the validity of it.

File: beacon-chain/p2p/service.go [hunk 2]
```diff
@@ -649,3 +654,23 @@ func logExternalDNSAddr(id peer.ID, addr string, port uint) {
...
```
Comment (prestonvanloon) [inline beacon-chain/p2p/service.go:659]: Please put this function in a more appropriate (or new) file. This file is already 600+ lines long.

Comment (fabdarice) [reply beacon-chain/p2p/service.go]: Moved to `p2p/utils.go`

File: beacon-chain/p2p/service.go (hunk 1 above)
Comment (fabdarice) [reply beacon-chain/p2p/service.go]: Hmm, unless I'm missing something, it'll be quite challenging to ensure that a PORT is open for UDP communications.
The updated current implementation (leveraging ListenUDP) will notify if an IP address isn't reachable, but I don't think, due to the nature of UDP (connectionless), we can detect if a specific port is open or not.

**Comment by fabdarice** [comment]
> Can you add a unit test?

I'll hold off until I ensure that I understand @nisdas  comment about checking UDP connectivity, but happy to write a unit tests once I get confirmation that I'm on the right track.

File: beacon-chain/p2p/service.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/p2p/service.go]: Not sure I understand, you can `Dial` the relevant IP + port to validate that there is a listening port running there. This works for both `TCP` and `UDP` . We have listening ports running for both protocols.

File: beacon-chain/p2p/utils.go [hunk 3]
```diff
@@ -131,3 +137,36 @@ func ipAddr() net.IP {
...
+			} else {
+				defer conn.Close()
+			}
+		case "udp":
+			addr, err := net.ResolveUDPAddr(protocol, a)
+			if err != nil {
+				log.Errorf("Could not resolve UDP address: %v", err)
+				return
+			}
+			_, err = net.ListenUDP("udp", addr)
```
Comment (nisdas) [inline beacon-chain/p2p/utils.go:163]: This won't work, you will be binding to a public ip which is not possible.

File: beacon-chain/p2p/utils.go [hunk 4]
```diff
@@ -131,3 +137,36 @@ func ipAddr() net.IP {
...
+		switch protocol {
+		case "tcp":
+			conn, err := net.DialTimeout(protocol, a, dialTimeout)
+			if err != nil {
+				log.WithFields(fields).Warn("IP address is not accessible")
+			} else {
+				defer conn.Close()
+			}
+		case "udp":
+			addr, err := net.ResolveUDPAddr(protocol, a)
```
Comment (nisdas) [inline beacon-chain/p2p/utils.go:158]: not sure why this is needed and is different from `tcp`

**Comment by fabdarice** [comment]
After chatting with @nisdas [chat](https://discordapp.com/channels/476244492043812875/667763844365680650/719759917963018341), it appears that `net.Dial` isn't properly validating if a connection to an address leveraging UDP is valid or not.

He suggested just doing TCP for now, for which I added a unit test as well.

**Pull request merged**